!!! FEATURE: Improve escr cli commands

mhsdesign opened the pull request:
<!-- 
    Thanks for your contribution, we appreciate it!

    The first section should explain briefly what is changed. 
    Some examples are always nice to showcase the use. 
    The content will be used in the change-logs and addresses 
    developers working with Neos.

    If there are issues regarding the topic of your PR link 
    them here as `related:` or `resolved:`
-->

Resolves: #4772

As discussed in the last weekly the following changes were made.

- `cr:replayAll` was renamed to `cr:projectionReplayAll`
- `cr:replay` was renamed to `cr:projectionReplay`
- `cr:projectionReplay` and `cr:projectionReplayAll` now both warn before continue and have a `--force` flag.
- `cr:prune` has  a `--force` flag.
- `cr:setup` has a `--reset-projections` flag.

Additionally `cr:projectionReplayAll` has a simple progress indicator.


**Upgrade instructions**

<!-- 
    Add upgrade instructions for breaking changes. 
    Explain who is affected, what has to be adjusted  
-->

**Review instructions**

<!-- 
    If your change is not explained fully by the first section you can 
    add more details here to help the reviewers understand the change.
    We have to understand what you did, why you did it and how we can 
    verify it works correctly and does no harm.
-->

**Checklist**

- [ ] Code follows the PSR-2 coding style
- [ ] Tests have been created, run and adjusted as needed
- [ ] The PR is created against the [lowest maintained branch](https://www.neos.io/features/release-roadmap.html)
- [ ] Reviewer - PR Title is brief but complete and starts with `FEATURE|TASK|BUGFIX`
- [ ] Reviewer - The first section explains the change briefly for change-logs
- [ ] Reviewer - Breaking Changes are marked with `!!!` and have upgrade-instructions

mhsdesign committed:
FEATURE: `flow cr:resetAll`

mhsdesign committed:
FEATURE: `cr:prune --assume-yes` flag to prune directly

and improve code of `cr:prune` by using `resetAllProjections` instead of `replayAll` as this was done before only implicit.

mhsdesign commented on Neos.ContentRepositoryRegistry/Classes/Command/CrCommandController.php:
```diff
@@ -144,15 +144,38 @@ public function replayAllCommand(string $contentRepository = 'default', bool $qu
+     * @param string $contentRepository Identifier of the Content Repository instance to operate on
+     * @param bool $quiet If set only fatal errors are rendered to the output
+     */
+    public function resetAllCommand(string $contentRepository = 'default', bool $quiet = false): void
```
maybe for people not reading the comments it might be misleading to have `cr:prune` and `cr:resetAll` in the same namespace.

Maybe we should add the `Projection` suffix?

`cr:resetAllProjections`

`cr:replayAllProjections`

`cr:replayProjection`

or we add another command controller:

`projection:resetAll`

`projection:replayAll`

`projection:replay`?

ahaeslich replied:
I agree here. Actually I struggle to understand what `reset` really means. What does it do? Why and when should someone use this command?

TBH I would have expected a better description in the PR for that as it will get into the changelog.

mhsdesign replied:
Yes this pr is not to be merged as is :)

my proposed new command will reset all the projections. This is implicit the first step of replayAll.

The reasoning to include is are: that we have a command to prune everything `cr:prune` and as the cr api allows to throw away all projections (like a cache flush) we should also allow this.

That command will most likely only be used when `cr:setup` fails because there is data in the projections. Basically for a super clean setup. Which hopefully is never need in 9.0 stable but was for example for https://github.com/neos/neos-development-collection/pull/4791.

ahaeslich replied:
Then we probably should include this change in some way and update the PR description of #4791 afterwards 😉. As of now the description is misleading.

mhsdesign replied:
the _real_ hopefully not that present usecase past 9.0 would be `cr:setup --reset-projections`.
Maybe that would be the way to go?

But i also prefer those commands:

```
cr:resetAllProjections

cr:replayAllProjections

cr:replayProjection
```

instead of the implicit `cr:replayAll`

mhsdesign replied:
`cr:projectionReplayAll`

`cr:projectionReplay`

prefer `cr:setup --reset-projections` over `cr:projectionResetAll`

mhsdesign commented on Neos.ContentRepositoryRegistry/Classes/Command/CrCommandController.php:
```diff
@@ -144,15 +144,38 @@ public function replayAllCommand(string $contentRepository = 'default', bool $qu
      */
-    public function pruneCommand(string $contentRepository = 'default'): void
+    public function pruneCommand(string $contentRepository = 'default', bool $assumeYes = false): void
     {
-        if (!$this->output->askConfirmation(sprintf("This will prune your content repository \"%s\". Are you sure to proceed? (y/n) ", $contentRepository), false)) {
+        if (!$assumeYes && !$this->output->askConfirmation(sprintf("This will prune your content repository \"%s\". Are you sure to proceed? (y/n) ", $contentRepository), false)) {
```
force and also add this flag to all other commands

mhsdesign committed:
TASK: Rename cli command to `flow cr:projectionReplayAll`

as suggested and discussed by christian

mhsdesign committed:
TASK: Add --force flag to cr cli commands

mhsdesign committed:
TASK: Add `flow cr:setup --reset-projections` instead of resetAll

mhsdesign committed:
TASK: Use `ProjectionReplayService` directly

mhsdesign committed:
TASK: Use `ProjectionReplayService::resetAllProjections` instead

as a relay all will just to that at first, but then there is no event to catch up to.
Reset is more obvious that way.

mhsdesign committed:
TASK: Simple progress bar for `flow cr:projectionReplayAll`

mhsdesign commented:
`cr:projectionResetAll` with confirmation and progress bar:

```
/home/neos: flow cr:projectionReplayAll          
> This will replay all projections in "default", which may take some time. Are you sure to proceed? (y/n) y
Replaying events for all projections of Content Repository "default" ...
 * replaying contentStream projection
 * replaying workspace projection
 * replaying nodeHiddenState projection
 * replaying documentUriPath projection
 * replaying change projection
 * replaying contentGraph projection
 * replaying assetUsage projection
Done.
```

`--reset-projections` flag:

```
/home/neos: flow cr:setup --reset-projections    
> Advanced Mode. The flag --reset-projections will reset all projections in "default", which leaves you with empty projections to be replayed. Are you sure to proceed? (y/n) y
All projections of Content Repository "default" were resettet.
Content Repository "default" was set up
```

And i noticed one other thing.

`cr:projectionReplayAll` seems much faster than replaying the contentGraph projection via `flow cr:projectionReplay contentGraph`. How can that beeeeeee?
-> that seems to be only an issue on my system? https://neos-project.slack.com/archives/C04PYL8H3/p1707730652428639

mhsdesign committed:
TASK: Improve and document cr:status

* adds `ContentRepositoryStatus::isOk` for simplicity

ahaeslich commented on Neos.ContentRepositoryRegistry/Classes/Command/CrCommandController.php:
```diff
@@ -33,18 +32,36 @@ public function __construct(
         $contentRepositoryId = ContentRepositoryId::fromString($contentRepository);
+
+        if ($resetProjections) {
+            if (!$this->output->askConfirmation(sprintf('> Advanced Mode. The flag --reset-projections will reset all projections in "%s", which leaves you with empty projections to be replayed. Are you sure to proceed? (y/n) ', $contentRepositoryId->value), false)) {
```
As `cr:setup` can be run in the CI didn't we also expect a `--force` option here 🤔?

mhsdesign replied:
Yes and no. The advanced mode `--reset-projections` is NOT expected to be run on ci. It should in general be not used. Only by us.

So a normal `cr:setup` will do just fine as before.
